File: cmd/atomsciflow_post.cpp
/************************************************************************
MIT License

Copyright (c) 2021 Deqi Tang

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
************************************************************************/


#include <boost/program_options.hpp>
//#include <filesystem>
#include <boost/filesystem.hpp>
#include <iostream>
#include <string>
#include <regex>
#include <cstdlib>
#include "atomsciflow/parser/cif.h"
#include "atomsciflow/parser/xyz.h"
#include "atomsciflow/parser/tools.h"
#include "atomsciflow/base/crystal.h"
#include "atomsciflow/utils.h"
#include "atomsciflow/abinit/abinit.h"

#include "atomsciflow_post_vasp.h"
#include "atomsciflow/cmd/cmd_utils.h"

// needs: libboost-dev, libboost-program-options-dev
namespace po = boost::program_options;


//namespace filesys = std::filesystem; 
namespace filesys = boost::filesystem;     // --std=c++11 -lboost_filesystem -lboost_system


int main(int argc, char const* argv[]) {
    //
    po::options_description global("Global options");
    global.add_options()
        ("subcommand", po::value<std::string>(), "command to execute")
        ("subargs", po::value<std::vector<std::string> >(), "Arguments for command")
        ("help, h", "print out help information");
        
    po::positional_options_description pos;
    pos.add("subcommand", 1).add("subargs", -1);
    
    po::variables_map vm;
    
    po::parsed_options parsed = po::command_line_parser(argc, argv).
        options(global).
        positional(pos).
        allow_unregistered().
        run();
        
    po::store(parsed, vm);
    
    log_cmd_start("atomsciflow-post.x");
    
    if (0 == vm.count("command")) { // or by vm.empty()
        std::cout << "You didn't specify any subcommand!\n";
        std::exit(1);
    }
    std::string subcommand = vm["subcommand"].as<std::string>();
    

    if ("cp2k" == subcommand) {

        
    } else if ("vasp" == subcommand) {
         //
        log_sub_cmd_start(subcommand);
        
        atomsciflow_post_vasp(parsed, vm);
        
        log_sub_cmd_end(subcommand); 

    } else if ("qe" == subcommand) {
        //

    } else {
        std::cout << "The specified subcommand is not defined!\n";
    }
        
    return 0;
}

File: atomsciflow/cmd/atomsciflow_post.py
"""
MIT License

Copyright (c) 2021 Deqi Tang

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
"""

import os
import sys
import argparse

from atomsciflow.cmd.cmd_utils import log_cmd_start, log_sub_cmd_start, log_sub_cmd_end
from atomsciflow.cmd.atomsciflow_post_cp2k import add_cp2k_post_subparser, cp2k_post_processor
from atomsciflow.cmd.atomsciflow_post_qe import add_qe_post_subparser, qe_post_processor
from atomsciflow.cmd.atomsciflow_post_siesta import add_siesta_post_subparser, siesta_post_processor
from atomsciflow.cmd.atomsciflow_post_vasp import add_vasp_post_subparser, vasp_post_processor
from atomsciflow.cmd.atomsciflow_post_abinit import add_abinit_post_subparser, abinit_post_processor
from atomsciflow.cmd.atomsciflow_post_octopus import add_octopus_post_subparser, octopus_post_processor
from atomsciflow.cmd.atomsciflow_post_elk import add_elk_post_subparser, elk_post_processor
from atomsciflow.cmd.atomsciflow_post_nwchem import add_nwchem_post_subparser, nwchem_post_processor
from atomsciflow.cmd.atomsciflow_post_gamessus import add_gamessus_post_subparser, gamessus_post_processor

def main():
    parser = argparse.ArgumentParser()
    subparsers = parser.add_subparsers(dest="subcommand", title="subcommand", description="choose one and only one subcommand")
    
    add_cp2k_post_subparser(subparsers)
    add_qe_post_subparser(subparsers)
    add_siesta_post_subparser(subparsers)
    add_vasp_post_subparser(subparsers)
    add_abinit_post_subparser(subparsers)
    add_octopus_post_subparser(subparsers)
    add_elk_post_subparser(subparsers)
    add_nwchem_post_subparser(subparsers)
    add_gamessus_post_subparser(subparsers)
    #
    args = parser.parse_args()

    if 1 == len(sys.argv):
        parser.print_help()
        sys.exit(1)        

    log_cmd_start("atomsciflow-post")

    if args.subcommand == "gamessus":
        log_sub_cmd_start(args.subcommand)
        gamessus_post_processor(args)
        log_sub_cmd_end(args.subcommand)  
    if args.subcommand == "vasp":
        log_sub_cmd_start(args.subcommand)
        vasp_post_processor(args)
        log_sub_cmd_end(args.subcommand)
    if args.subcommand == "cp2k":
        log_sub_cmd_start(args.subcommand)
        cp2k_post_processor(args)
        log_sub_cmd_end(args.subcommand)
    if args.subcommand == "abinit":
        log_sub_cmd_start(args.subcommand)
        abinit_post_processor(args)
        log_sub_cmd_end(args.subcommand)
    if args.subcommand == "elk":
        log_sub_cmd_start(args.subcommand)
        elk_post_processor(args)
        log_sub_cmd_end(args.subcommand)
    if args.subcommand == "qe":
        log_sub_cmd_start(args.subcommand)
        qe_post_processor(args)
        log_sub_cmd_end(args.subcommand)
    if args.subcommand == "qmcpack":
        pass
    if args.subcommand == "siesta":
        log_sub_cmd_start(args.subcommand)
        siesta_post_processor(args)
        log_sub_cmd_end(args.subcommand)
    if args.subcommand == "dalton":
        pass                                     
    if args.subcommand == "octopus":
        log_sub_cmd_start(args.subcommand)
        octopus_post_processor(args)
        log_sub_cmd_end(args.subcommand)                                
    if args.subcommand == "nwchem":
        log_sub_cmd_start(args.subcommand)
        nwchem_post_processor(args)
        log_sub_cmd_end(args.subcommand)                                  
    if args.subcommand == "orca":
        pass                                      
    if args.subcommand == "gaussian":
        pass                                      
    if args.subcommand == "qchem":
        pass                                     
    if args.subcommand == "exciting":
        pass                                      
    if args.subcommand == "xtb":
        pass                                      
    if args.subcommand == "gromacs":
        pass                                  
    if args.subcommand == "lammps":
        pass                                   
    if args.subcommand == "dftbplus":
        pass

if __name__ == "__main__":
    main()
